Components/Windows/configure-wsus-server/configure-wsus-server.yaml

```diff
@@ -0,0 +1,113 @@
+# SPDX-License-Identifier: MIT-0
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this
+# software and associated documentation files (the "Software"), to deal in the Software
+# without restriction, including without limitation the rights to use, copy, modify,
+# merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+# INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+# PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+name: WsusConfiguration
+description: 'Configure WSUS server via registry keys'
+schemaVersion: 1.0
+constants:
+  - WUServer:
+      type: string
+      value: http://10.0.0.1:8530
+  - WUStatusServer:
+      type: string
+      value: http://10.0.0.1:8530
+  - TargetGroup:
+      type: string
+      value: Ec2ImageBuilder
+phases:
+  - name: build
+    steps:
+      - name: ConfigureWsusServer
+        action: ExecutePowerShell
+        inputs:
+          commands:
+            - |
+              $ErrorActionPreference = 'Stop'
+              $ProgressPreference = 'SilentlyContinue'
+              
+              $registryRoot = "HKLM:\Software\Policies\Microsoft\Windows\WindowsUpdate"
+              $registryRootAu = $registryRoot + "\AU"
+
+              Write-Host "Setting Windows Update Server..."
+              Set-ItemProperty -Path $registryRoot -Name "WUServer" -Type String -Value "{{ WUServer }}"
+              Write-Host "done."
+              
+              Write-Host "Setting Windows Update Status Server..."
+              Set-ItemProperty -Path $registryRoot -Name "WUStatusServer" -Type String -Value "{{ WUStatusServer }}"
+              Write-Host "done."
+              
+              Write-Host "Enabling client-side targeting..."
+              Set-ItemProperty -Path $registryRoot -Name "TargetGroupEnabled" -Type DWord -Value 1
+              Write-Host "done."
+              
+              Write-Host "Setting Windows Update TargetGroup..."
+              Set-ItemProperty -Path $registryRoot -Name "TargetGroup" -Type String -Value "{{ TargetGroup }}"
+              Write-Host "done."
+              
+              Write-Host "Switching update source from Windows Update to WSUS Server..."
+              Set-ItemProperty -Path $registryRootAu -Name "UseWUServer" -Type DWord -Value 1
+              Write-Host "done."
```

Comment:
I wanted to gather thoughts on using [ExecutePowerShell](https://docs.aws.amazon.com/imagebuilder/latest/userguide/image-builder-action-modules.html#image-builder-action-modules-executepowershell) versus the [SetRegistry](https://docs.aws.amazon.com/imagebuilder/latest/userguide/image-builder-action-modules.html#image-builder-action-modules-setregistry) action module here.

I'm ok with this using PowerShell, but I wonder if perhaps it should use SetRegistry, to help expand knowledge that the action module exists rather than defaulting to a script.

It does make the component longer (in terms on lines-of-code), but the intent of SetRegistry was to help remove the need of customers needing to author PowerShell for configuring the Registry.

Reply:
Hi Andrew,

I was thinkink about using SetRegistry but decided to use ExecutePowerShell for future expansions:

As there is no mechanism to support conditions in Action modules the PowerShell Variant is more flexible. I plan to extend the document to not only support client-side targetting, which would require branching based on the input values.